### DISCUSSION TITLE
RLS enabled for table, not blocking SELECTs

### DISCUSSION THREAD

**Discussion by epd** [supabase/supabase#17796] [Questions] [answered]
Hi there - I am having some issues enforcing RLS on reads. I have RLS enabled on a table `accounts_table` with no policies defined. I am using a browser client with a combination of the anon key and an authenticated session (I can see this in the network requests being made both locally and to the SaaS Supabase instance). I have also decoded the JWTs to ensure the `role` field is `anon` and `authenticated` respectively.

My understanding from https://www.postgresql.org/docs/current/ddl-rowsecurity.html#:~:text=If%20no%20policy%20exists%20for,to%20roles%2C%20or%20to%20both. is that the default policy for tables with RLS enabled is `deny`.

Using the browser client, I am able to verify that `INSERTS` are blocked with a 403 (due to RLS). However, selecting from this table yields all the data despite having no permissive RLS policies. I have confirmed this behavior on both local dev (running the Docker stack) as well as our production instances (Vercel + Supabase).

I would expect this to return a similar 403 on `SELECT` operations, despite being authenticated or using the anon key.

Here are a few queries to show the state of the DB:
<img width="921" alt="Screenshot 2023-09-28 at 6 13 52 PM" src="https://github.com/supabase/supabase/assets/412465/6b929e9b-ebe7-4ad9-8755-063c7df7eb9a">

From the above, you can see RLS is enabled for the table.

<img width="924" alt="Screenshot 2023-09-28 at 6 14 14 PM" src="https://github.com/supabase/supabase/assets/412465/f56b1ca7-4c26-4e79-b4a4-98e145e8905a">

From the above, you can see there are no policies for the table.

Here is some sample code, where `supabase` is initialized from the `@supabase/auth-helpers-nextjs` package, which I have also recently updated both this package and the official SDK to `@latest`:
<img width="731" alt="Screenshot 2023-09-28 at 6 25 39 PM" src="https://github.com/supabase/supabase/assets/412465/ff2e72cf-e595-4c7c-9d4f-a86082d3f9dc">
<img width="817" alt="Screenshot 2023-09-28 at 6 25 11 PM" src="https://github.com/supabase/supabase/assets/412465/8af0364e-f8f0-4a74-9294-59f370bbfcfa">

I have also verified there are no unusual/unexpected Postgres logs locally or remote and that both roles are not granted to bypass RLS:
<img width="829" alt="Screenshot 2023-09-28 at 6 30 17 PM" src="https://github.com/supabase/supabase/assets/412465/edd338a0-3cf2-4389-a7c0-f89e208e1510">

I have done a search for similar discussions and have only found this Stack Overflow post, which did not yield much information: https://stackoverflow.com/questions/71072815/supabase-rls-not-restricting-reads.

Any ideas of what I could be doing wrong here? Thank you kindly in advance 🙏

**Comment by GaryAustin1** [comment]
First select not meeting RLS does not error, it returns empty array with no error.

My guess is you are using next.js 13 and it is caching your select from a previous operation before you enabled RLS or dropped a select policy maybe?

You could test by changing a column in your select so the URL is different, if that works then Next.js is caching your selects.

Edit: One user in Supabase's Discord used this to disable the caching.. `export const fetchCache = 'force-no-store'` .

**Reply by epd** [reply]
hey @GaryAustin1 - thanks a lot for the reply and ideas, it is much appreciated 🙏 

I did everything I could to bust the cache, but I'm still seeing the same results (`force-no-store`, `force-dynamic`, rename the route, change the data that is returned, remove `.next` directory, etc.). I can also see a 200 coming from the REST API with cache control headers properly set in the browser (as this is a client component). I'll keep on debugging, maybe it is something with how I am configuring/using the browser client and the Next.js helpers.

I'll report back if I make any progress and maybe join the Discord as well.

**Reply by epd** [reply, marked as answer]
@GaryAustin1 I figured it out, thanks to your response here: https://github.com/orgs/supabase/discussions/17692#discussioncomment-7103587

I erroneously created this view without `security_invoker = true`. Thank you for all your help.
